### PR TITLE
Filter by fasta

### PR DESCRIPTION
**Pull request by pravs3683**
Added a feature where id list can be accession number in the fasta header (necessarily not the complete fasta header). 
Eg. filter fasta id based on Uniprot accession number from a Uniprot like fasta file. 

This modification uses pattern recognition to match the accessions in the header.

### CODE REVIEW

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.xml [hunk 1]
```diff
@@ -51,6 +54,24 @@ $dedup
...
+                
+                
+                <conditional name="id_regex">
+                    <param name="find" type="select" label="Match IDs by">
+                        <option value="beginning">ID is expected at the beginning: &gt;ID</option>
+                        <option value="pattern">regex pattern</option>
+                    </param>
+                    <when value="beginning"/>
+                    <when value="pattern">
+                        <param name="pattern" type="text" value="" label="regex search pattern for ID">
```
Comment (bgruening) [inline tools/filter_by_fasta_ids/filter_by_fasta_ids.xml]: The default value in the python script is `>([^| ]+)` isn't it? Maybe we should add this here or mention it in the help of this param.

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.py [hunk 2]
```diff
@@ -64,14 +63,21 @@ def main():
     parser.add_argument('-d', help='Path to discarded entries file')
     header_criteria = parser.add_mutually_exclusive_group()
     header_criteria.add_argument('--id_list', help='Path to the ID list file')
+    parser.add_argument('--pattern', default=None, help='regex earch attern for ID in Fasta entry')
```
Comment (bgruening) [inline tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: "search pattern" ... mention the implicit default value as well?

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.py [hunk 3]
```diff
@@ -64,14 +63,21 @@ def main():
...
     sequence_criteria.add_argument('--sequence_regexp', help='Regular expression pattern the header should match')
     parser.add_argument('--max_length', type=int, help='Maximum sequence length')
     parser.add_argument('--dedup', action='store_true', default=False, help='Whether to remove duplicate sequences')
     options = parser.parse_args()
-
+    
+    
+    pattern =  options.pattern if options.pattern else '>([^| ]+)'
+    if not re.match('^.*[(](?![?]:).*[)].*$',pattern):
+        print('pattern: "%s" did not include "%%s"' % pattern)
```
Comment (bgruening) [inline tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: is %% really want you want?

Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: Yes.

Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: We want user to specify ( ) in the regex for capturing group which can be used to pass in the string to match the fasta header.

Comment (bgruening) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: I mean in the string substitution :)

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.xml (hunk 1 above)
Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.xml]: Modified according to your suggestion

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.py (hunk 2 above)
Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: Modified

File: tools/filter_by_fasta_ids/filter_by_fasta_ids.py (hunk 3 above)
Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: How about if I change it to 

print('pattern: "%s" did not include capture group "()" in regex ' % pattern)

Comment (bgruening) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: sure

Comment (pravs3683) [reply tools/filter_by_fasta_ids/filter_by_fasta_ids.py]: Done!! Thanks a lot!!

**Comment by bgruening** [comment]
Thanks!

**Pull request merged**